### DISCUSSION TITLE
Early stopping in individual trials for reinforcement learning with Stable Baselines 3

### DISCUSSION THREAD

**Discussion by Anth0rx** [optuna/optuna#4495] [Q&A]
In optunas example on RL it implements a [`TrialEvalCallback`](https://github.com/optuna/optuna-examples/blob/main/rl/sb3_simple.py#L78) class which inherits from stable-baselines3's  [`EvalCallback`](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/callbacks.py#L326) class.

Passing the `callback_after_eval` argument with [`StopTrainingOnNoModelImprovement`](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/callbacks.py#L620) in the `super().__init__()` block does not stop the trial early, letting it run for the whole `N_TIMESTEPS`.

Does anyone have a working Stable Baselines 3 example on how to early stop a trial for which the current model is not improving anymore?

**Comment by nzw0301** [comment]
Could you share the minimal reproducible code with us?

**Reply by Anth0rx** [reply]
Sure, I made the following changes to your [`sb3_example.py`](https://github.com/optuna/optuna-examples/blob/main/rl/sb3_simple.py):

1. Import `StopTrainingOnNoModelImprovement`:
```python
from stable_baselines3.common.callbacks import EvalCallback, StopTrainingOnNoModelImprovement 
```

2. Increase `N_EVALUATIONS`:
```python
N_EVALUATIONS = 10
```

3. Enable verbose output:
```python
DEFAULT_HYPERPARAMS = {
    "policy": "MlpPolicy",
    "env": ENV_ID,
    "verbose": 1,
}
```

4. Add `StopTrainingOnNoModelImprovement` callback to `super().__init__()`:
```python
super().__init__(
            eval_env=eval_env,
            n_eval_episodes=n_eval_episodes,
            eval_freq=eval_freq,
            deterministic=deterministic,
            verbose=verbose,
            callback_after_eval=StopTrainingOnNoModelImprovement(max_no_improvement_evals=2, min_evals=0, verbose=1)
        )
```

### Here is my final script:
```python
""" Optuna example that optimizes the hyperparameters of
a reinforcement learning agent using A2C implementation from Stable-Baselines3
on an OpenAI Gym environment.

This is a simplified version of what can be found in https://github.com/DLR-RM/rl-baselines3-zoo.

You can run this example as follows:
    $ python sb3_simple.py

"""
from typing import Any
from typing import Dict

import gym
import optuna
from optuna.pruners import MedianPruner
from optuna.samplers import TPESampler
from stable_baselines3 import A2C
from stable_baselines3.common.callbacks import EvalCallback, StopTrainingOnNoModelImprovement
from stable_baselines3.common.monitor import Monitor
import torch
import torch.nn as nn


N_TRIALS = 100
N_STARTUP_TRIALS = 5
N_EVALUATIONS = 10
N_TIMESTEPS = int(2e4)
EVAL_FREQ = int(N_TIMESTEPS / N_EVALUATIONS)
N_EVAL_EPISODES = 3

ENV_ID = "CartPole-v1"

DEFAULT_HYPERPARAMS = {
    "policy": "MlpPolicy",
    "env": ENV_ID,
    "verbose": 1,
}


def sample_a2c_params(trial: optuna.Trial) -> Dict[str, Any]:
    """Sampler for A2C hyperparameters."""
    gamma = 1.0 - trial.suggest_float("gamma", 0.0001, 0.1, log=True)
    max_grad_norm = trial.suggest_float("max_grad_norm", 0.3, 5.0, log=True)
    gae_lambda = 1.0 - trial.suggest_float("gae_lambda", 0.001, 0.2, log=True)
    n_steps = 2 ** trial.suggest_int("exponent_n_steps", 3, 10)
    learning_rate = trial.suggest_float("lr", 1e-5, 1, log=True)
    ent_coef = trial.suggest_float("ent_coef", 0.00000001, 0.1, log=True)
    ortho_init = trial.suggest_categorical("ortho_init", [False, True])
    net_arch = trial.suggest_categorical("net_arch", ["tiny", "small"])
    activation_fn = trial.suggest_categorical("activation_fn", ["tanh", "relu"])

    # Display true values.
    trial.set_user_attr("gamma_", gamma)
    trial.set_user_attr("gae_lambda_", gae_lambda)
    trial.set_user_attr("n_steps", n_steps)

    net_arch = [
        {"pi": [64], "vf": [64]} if net_arch == "tiny" else {"pi": [64, 64], "vf": [64, 64]}
    ]

    activation_fn = {"tanh": nn.Tanh, "relu": nn.ReLU}[activation_fn]

    return {
        "n_steps": n_steps,
        "gamma": gamma,
        "gae_lambda": gae_lambda,
        "learning_rate": learning_rate,
        "ent_coef": ent_coef,
        "max_grad_norm": max_grad_norm,
        "policy_kwargs": {
            "net_arch": net_arch,
            "activation_fn": activation_fn,
            "ortho_init": ortho_init,
        },
    }


class TrialEvalCallback(EvalCallback):
    """Callback used for evaluating and reporting a trial."""

    def __init__(
        self,
        eval_env: gym.Env,
        trial: optuna.Trial,
        n_eval_episodes: int = 5,
        eval_freq: int = 10000,
        deterministic: bool = True,
        verbose: int = 0,
    ):
        super().__init__(
            eval_env=eval_env,
            n_eval_episodes=n_eval_episodes,
            eval_freq=eval_freq,
            deterministic=deterministic,
            verbose=verbose,
            callback_after_eval=StopTrainingOnNoModelImprovement(max_no_improvement_evals=2, min_evals=0, verbose=1)
        )
        self.trial = trial
        self.eval_idx = 0
        self.is_pruned = False

    def _on_step(self) -> bool:
        if self.eval_freq > 0 and self.n_calls % self.eval_freq == 0:
            super()._on_step()
            self.eval_idx += 1
            self.trial.report(self.last_mean_reward, self.eval_idx)
            # Prune trial if need.
            if self.trial.should_prune():
                self.is_pruned = True
                return False
        return True


def objective(trial: optuna.Trial) -> float:
    kwargs = DEFAULT_HYPERPARAMS.copy()
    # Sample hyperparameters.
    kwargs.update(sample_a2c_params(trial))
    # Create the RL model.
    model = A2C(**kwargs)
    # Create env used for evaluation.
    eval_env = Monitor(gym.make(ENV_ID)) ###
    # Create the callback that will periodically evaluate and report the performance.
    eval_callback = TrialEvalCallback(
        eval_env, trial, n_eval_episodes=N_EVAL_EPISODES, eval_freq=EVAL_FREQ, deterministic=True
    )

    nan_encountered = False
    try:
        model.learn(N_TIMESTEPS, callback=eval_callback)
    except AssertionError as e:
        # Sometimes, random hyperparams can generate NaN.
        print(e)
        nan_encountered = True
    finally:
        # Free memory.
        model.env.close()
        eval_env.close()

    # Tell the optimizer that the trial failed.
    if nan_encountered:
        return float("nan")

    if eval_callback.is_pruned:
        raise optuna.exceptions.TrialPruned()

    return eval_callback.last_mean_reward


if __name__ == "__main__":
    # Set pytorch num threads to 1 for faster training.
    torch.set_num_threads(1)

    sampler = TPESampler(n_startup_trials=N_STARTUP_TRIALS)
    # Do not prune before 1/3 of the max budget is used.
    pruner = MedianPruner(n_startup_trials=N_STARTUP_TRIALS, n_warmup_steps=N_EVALUATIONS // 3)

    study = optuna.create_study(sampler=sampler, pruner=pruner, direction="maximize")
    try:
        study.optimize(objective, n_trials=N_TRIALS, timeout=600)
    except KeyboardInterrupt:
        pass

    print("Number of finished trials: ", len(study.trials))

    print("Best trial:")
    trial = study.best_trial

    print("  Value: ", trial.value)

    print("  Params: ")
    for key, value in trial.params.items():
        print("    {}: {}".format(key, value))

    print("  User attrs:")
    for key, value in trial.user_attrs.items():
        print("    {}: {}".format(key, value))
```

### What I would expect:
The running trial should end if the `StopTrainingOnNoModelImprovement` callback is triggered. The verbose output shows the callback is triggered multiple times in sequence but it does not seem to end the trial:

```
Stopping training because there was no new best model in the last 3 evaluations
[...]
Stopping training because there was no new best model in the last 4 evaluations
[...]
Stopping training because there was no new best model in the last 5 evaluations
[...]
Stopping training because there was no new best model in the last 6 evaluations
[...]
Stopping training because there was no new best model in the last 7 evaluations
[...]
[I 2023-03-08 11:30:55,093] Trial 0 finished with value: [...]. Best is trial 0 with value: [...].
```

**Comment by KandBM** [comment]
I've written a pruner to miimic SB3's StopTrainingOnNoModelImprovement callback:
`class NoImprovementPruner(optuna.pruners.BasePruner): 
  
    def __init__(self, direction:str, warmup_steps:int=2,patience:int=2):
        """Prunes trials when the objective value stops improving
        only tested for single objectives
        direction: string of either maximize or minimize
        warmup speficies the number of evals before the trail can be pruned
        patience specifies the number of evals with no improvement before pruning"""
        self.warmup_steps = warmup_steps
        dir_map = {'minimize':1,'maximize':-1} #prune assumes a minimization
        assert direction in dir_map.keys(), f'Direction must be in {dir_map.keys()}'
        self.direction = dir_map[direction] #score coeff avoids rewriting for different directions
            #could direction be inferred from the study object in prune?
        self.patience = patience
        self.best_intermediate_result = self.direction*float('inf')
        self.best_step = 0


    def prune(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -> bool:
        # Get the latest intermediate result (the value at the current step)
        step = trial.last_step
        if step is None: #first step
            return False
        current_value = trial.intermediate_values[step]

        # During warmup, update the best intermediate result and do not prune
        if step < self.warmup_steps:
            self.best_intermediate_result = self.direction*min(self.direction*self.best_intermediate_result, 
                                                               self.direction*current_value)
            return False

        # After warmup, if the current value is better, update the best intermediate result and step
        if self.direction*current_value < self.direction*self.best_intermediate_result:
            self.best_intermediate_result = current_value
            self.best_step = step

        # Prune the trial if the number of evaluations without improvement exceeds the patience
        if step - self.best_step > self.patience:
            self.best_intermediate_result = self.direction*float('inf')
            self.best_step = 0
            return True

        return False`
